### PR TITLE
os syscalls: add getifaddrs to singleton

### PR DESCRIPTION
**Pull request by junr03**
Commit Message: os syscalls - add getifaddrs to singleton
Additional Description: this PR reverts #18566 and adjusts code in order to be able to compile against android in Envoy Mobile
Risk Level: low, new API. Also tested against existing code.
Testing: existing UT.

Signed-off-by: Jose Nino <jnino@lyft.com>

### CODE REVIEW

File: envoy/api/os_sys_calls.h [hunk 1]
```diff
@@ -17,6 +19,20 @@ struct EnvoyTcpInfo {
   std::chrono::microseconds tcpi_rtt;
 };
 
+// Small struct to avoid exposing ifaddrs -- which is not defined in all platforms -- to the
+// codebase.
+struct InterfaceAddress {
```
Comment (junr03) [inline envoy/api/os_sys_calls.h:24]: happy to move this to address.h if we think it makes more sense there. I couldn't decide.

File: source/common/api/posix/os_sys_calls_impl.cc [hunk 2]
```diff
@@ -282,5 +283,53 @@ SysCallBoolResult OsSysCallsImpl::socketTcpInfo([[maybe_unused]] os_fd_t sockfd,
   return {false, EOPNOTSUPP};
 }
 
+bool OsSysCallsImpl::supportsGetifaddrs() const {
+// TODO: eliminate this branching by upstreaming an alternative Android implementation
+// e.g.: https://github.com/envoyproxy/envoy-mobile/blob/main/third_party/android/ifaddrs-android.h
+#if defined(__ANDROID_API__) && __ANDROID_API__ < 24
```
Comment (junr03) [inline source/common/api/posix/os_sys_calls_impl.cc:289]: @goaway I attempted to coerce the Socket options pattern here. The problem is that the definition of the macro in Envoy Mobile still wouldn't be available to this code. Unless I am missing some macro/preprocessor magic. Any pointers would be helpful.

Comment (junr03) [reply source/common/api/posix/os_sys_calls_impl.cc:303]: Fixed by adding a setter for a lambda in the singleton.

**Comment by junr03** [comment]
@mattklein123 still not done updating. I will ping you when this is ready for review.

File: envoy/api/os_sys_calls.h [hunk 3]
```diff
@@ -17,6 +21,22 @@ struct EnvoyTcpInfo {
...
+  InterfaceAddress(absl::string_view ifa_name, unsigned int ifa_flags,
```
Comment (mattklein123) [inline envoy/api/os_sys_calls.h:27]: nit: can you just s/ifa/interface. It's easier to read.

File: source/common/api/posix/os_sys_calls_impl.cc [hunk 4]
```diff
@@ -296,5 +297,61 @@ SysCallBoolResult OsSysCallsImpl::socketTcpInfo([[maybe_unused]] os_fd_t sockfd,
...
+  if (alternate_getifaddrs_.has_value()) {
```
Comment (mattklein123) [inline source/common/api/posix/os_sys_calls_impl.cc:304]: Shouldn't you be doing this in all cases otherwise the interface docs are misleading?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM with small comments, thanks.

/wait

File: source/common/api/posix/os_sys_calls_impl.cc (hunk 4 above)
Comment (junr03) [reply source/common/api/posix/os_sys_calls_impl.cc:304]: @mattklein123 I think I might be misunderstanding your point. Where else should I be checking this that is not checked?

Comment (mattklein123) [reply source/common/api/posix/os_sys_calls_impl.cc:304]: You should check this outside of the #ifdef ?

Comment (junr03) [reply source/common/api/posix/os_sys_calls_impl.cc:304]: I see what you mean now @mattklein123. My thought was that in posix the class would default to true for _supporting_ getifaddrs regardless of if an alternate implementation is given. Only for Android<24 do we need to check for existence of an alternate implementation.

Comment (mattklein123) [reply source/common/api/posix/os_sys_calls_impl.cc:304]: OK I see. Makes sense. Can you just add a small comment?

/wait

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**